flexget/plugins/clients/aria2.py

```diff
@@ -83,36 +87,76 @@ def on_task_output(self, task, config):
-            except xmlrpc.client.Fault as err:
-                log.critical('Fault code %s message %s', err.faultCode, err.faultString)
-                entry.fail('Aria2 communication Fault')
-            except Exception as e:
-                log.debug('Exception type %s', type(e), exc_info=True)
-                raise
+            # check for content_files first, then use url or title if not
+            if 'content_files' not in entry:
+                if entry['url']:
+                    entry['content_files'] = [entry['url']]
+                else:
+                    entry['content_files'] = [entry['title']]
+            else:
+                if not isinstance(entry['content_files'], list):
+                    entry['content_files'] = [entry['content_files']]
+            for current_file in entry['content_files']:
+                # expose current filename for rendering purposes
+                entry['current_file'] = current_file
+                if task.options.test:
+                    log.verbose('Would add `%s` to aria2.', entry['current_file'])
+                    continue
+                try:
+                    self.add_entry(aria2, entry, config)
+                except socket_error as se:
+                    entry.fail('Unable to reach Aria2: %s', se)
+                except xmlrpc.client.Fault as err:
+                    log.critical('Fault code %s message %s', err.faultCode, err.faultString)
+                    entry.fail('Aria2 communication Fault')
+                except Exception as e:
+                    log.debug('Exception type %s', type(e), exc_info=True)
+                    raise
+                else:
+                    log.debug('Added to aria2: `%s`', entry['current_file'])
 
     def add_entry(self, aria2, entry, config):
         """
         Add entry to Aria2
         """
+        # reset every loop or it won't work correctly after the first
         options = config['options']
         try:
             options['dir'] = os.path.expanduser(entry.render(config['path']).rstrip('/'))
         except RenderError as e:
             entry.fail('failed to render \'path\': %s' % e)
             return
+        if config['keep_structure']:
+            # TODO: consider case where current_file is a URL using urlparse,
+            #       to strip out protocol & hostname
+            options['dir'] = os.path.join(options['dir'], os.path.dirname(entry['current_file']))
         if 'filename' in config:
-             try:
-                 options['out'] = os.path.expanduser(entry.render(config['filename']))
-             except RenderError as e:
-                 entry.fail('failed to render \'filename\': %s' % e)
-                 return
+            try:
+                options['out'] = os.path.expanduser(entry.render(config['filename']))
+            except RenderError as e:
+                entry.fail('failed to render \'filename\': %s' % e)
+                return
```

Comment:
I'm not sure just returning is the best option on failure, as it doesn't look you implicitly handle that